README.md
## SageMaker SparkMagic Library

[![Version](https://img.shields.io/pypi/v/sagemaker-studio-sparkmagic-lib.svg)](https://pypi.org/project/sagemaker-studio-sparkmagic-lib/)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)


This is a CLI tool for generating configuration of SparkMagic, Kerberos required  to connect to EMR cluster. In particular, it generates following two files

1. **SparkMagic Config**: This config file contains information needed to connect SparkMagic kernel's running on studio to Livy application running on EMR. CLI obtains EMR cluster details like Ip address etc.. by describing EMR cluster

2. **Krb5.conf**: If EMR cluster uses kerberos security configuration, this library also generates krb5.conf needed for user authentication on studio 

### Usage

This CLI tool comes pre-installed on Studio SparkMagic Image. It can be used from any notebook created from that image. 

#### Connecting to non-kerberos cluster: 
In a notebook cell, execute following commands

```
%local

!sm-sparkmagic connect --cluster-id "j-xxxxxxxxx"
```

sample output:

```
Successfully read emr cluster(j-xxxxxxxx) details
SparkMagic config file is written to location /etc/sparkmagic/config.json
Completed setting up configuration files for SparkMagic to connect to EMR cluster j-xxxxxxxx


Please complete following steps to complete the connection
1. Restart kernel to complete your setup. This is required so SparkMagic can pickup generated configuration
```

#### Connecting to kerberos cluster: 

It's very similar to non-kerberos cluster, except you can pass 

```
!sm-sparkmagic connect --cluster-id "j-xxxxxxxx" --user-name "ec2-user"
```

sample output:

```
Please follow below steps to complete the setup:
1. Please open image terminal and run 'kinit ec2-user'(user_name: ec2-user) to get kerberos ticket
2. Restart kernel to complete your setup. This is required so SparkMagic can pickup generated configuration
```

#### Connecting to EMR cluster in another account 
To setup configuration for EMR cluster in another account, run following command

```
%local

!sm-sparkmagic connect --cluster-id "j-xxxxx" --role-arn "arn:aws:iam::222222222222:role/role-on-emr-cluster-account"
```

### FAQ
* Can I connect to multiple clusters at same time?
  * You can only connect to one cluster at a time. Tool generates configuration needed to connect to one cluster. If you want to connect to different cluster, one has to re-execute the command providing different cell
* Can I use this CLI on non-SparkMagic image on studio?
  * This cli only comes pre-installed on SparkMagic Image. One can install on any other image if needed
* Can I use this library on SageMaker Notebook instances?
  * It does not come installed on Notebooks either, but you can install and try using it. You may have to relocate SparkMagic conf file
  

### Installing
Install the CLI using pip. 

```
pip install sagemaker-studio-sparkmagic-lib
```

Following extra permissions are required on the role to be able to describe cluster
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "elasticmapreduce:DescribeCluster",
                "elasticmapreduce:DescribeSecurityConfiguration",
                "elasticmapreduce:ListInstances"
            ],
            "Resource": "arn:aws:elasticmapreduce:*:*:cluster/*"
        }
    ]
}
```

### Development

* checkout the repository, and install locally

```
make install
```

* To test locally, you can start python3 REPL and run following python code

```python
import sagemaker_studio_sparkmagic_lib.sparkmagic as sm
sm.connect_to_emr_cluster(cluster_id= "j-xxx", user_name="ec2-user", krb_file_override_path="/tmp/krb5.conf",
     spark_magic_override_path="/tmp/config.json", restart_kernel=False)
```

* To test on studio, create a tar ball and install on studio or your custom image accordingly

```
python setup.py sdist
```
## Security

See [CONTRIBUTING](CONTRIBUTING.md#security-issue-notifications) for more information.

## License

This library is licensed under the MIT-0 License. See the LICENSE file.




CHANGELOG.md
# Changelog

## 0.1.4
*  Allow surfacing of primary node public dns

## 0.1.3
* Paginate list_instances call

## 0.1.2
* Fix "krb5.conf" generated file for external KDC cases
* Add ChangeLog and Update README


sagemaker_studio_sparkmagic_lib/tests/emr_test.py
from unittest.mock import patch

import boto3
import pytest
from botocore.stub import Stubber

from sagemaker_studio_sparkmagic_lib.emr import EMRCluster


def test_get_cluster_happy_case_non_kerberos():
    emr = boto3.client("emr", region_name="us-west-2")
    with Stubber(emr) as emr_stub:
        describe_cluster_response = {
            "Cluster": {
                "Id": "j-3DD9ZR01DAU14",
                "Name": "Mycluster",
                "MasterPublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
            }
        }
        list_instances_response = {
            "Instances": [
                {
                    "Id": "j-3DD9ZR01DAU14",
                    "Ec2InstanceId": "i-0736242069217a485",
                    "PublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
                    "PublicIpAddress": "34.222.47.14",
                    "PrivateDnsName": "ip-172-31-1-113.us-west-2.compute.internal",
                    "PrivateIpAddress": "172.31.1.113",
                }
            ]
        }
        emr_stub.add_response("describe_cluster", describe_cluster_response)
        emr_stub.add_response("list_instances", list_instances_response)
        emr_cluster = EMRCluster(cluster_id="j-3DD9ZR01DAU14", emr=emr)

        assert emr_cluster._cluster == describe_cluster_response["Cluster"]
        assert emr_cluster._instances == list_instances_response["Instances"]
        assert emr_cluster._sec_conf is None
        assert not emr_cluster.is_krb_cluster


def test_get_private_cluster_happy_case_non_kerberos():
    emr = boto3.client("emr", region_name="us-west-2")
    with Stubber(emr) as emr_stub:
        describe_cluster_response = {
            "Cluster": {
                "Id": "j-3DD9ZR01DAU14",
                "Name": "Mycluster",
                "MasterPublicDnsName": "ip-172-31-1-113.us-west-2.compute.internal",
            }
        }
        list_instances_response = {
            "Instances": [
                {
                    "Id": "j-3DD9ZR01DAU14",
                    "Ec2InstanceId": "i-0736242069217a485",
                    "PublicDnsName": "",
                    "PrivateDnsName": "ip-172-31-1-113.us-west-2.compute.internal",
                    "PrivateIpAddress": "172.31.1.113",
                }
            ]
        }
        emr_stub.add_response("describe_cluster", describe_cluster_response)
        emr_stub.add_response("list_instances", list_instances_response)
        emr_cluster = EMRCluster(cluster_id="j-3DD9ZR01DAU14", emr=emr)

        assert emr_cluster._cluster == describe_cluster_response["Cluster"]
        assert emr_cluster._instances == list_instances_response["Instances"]
        assert emr_cluster._sec_conf is None
        assert not emr_cluster.is_krb_cluster
        assert (
            emr_cluster.primary_node_private_dns_name()
            == "ip-172-31-1-113.us-west-2.compute.internal"
        )
        assert emr_cluster.primary_node_public_dns_name() == ""


def test_get_cluster_happy_case_non_kerberos_with_pagination():
    emr = boto3.client("emr", region_name="us-west-2")
    with Stubber(emr) as emr_stub:
        describe_cluster_response = {
            "Cluster": {
                "Id": "j-3DD9ZR01DAU14",
                "Name": "Mycluster",
                "MasterPublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
            }
        }
        list_instances_response_page_1 = {
            "Instances": [
                {
                    "Id": "j-3DD9ZR01DAU14",
                    "Ec2InstanceId": "i-0736242069217a485",
                    "PublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
                    "PublicIpAddress": "34.222.47.14",
                    "PrivateDnsName": "ip-172-31-1-113.us-west-2.compute.internal",
                    "PrivateIpAddress": "172.31.1.113",
                }
            ],
            "Marker": "nextToken",
        }
        list_instances_response_page_2 = {
            "Instances": [
                {
                    "Id": "j-3DD9ZR01DAU14",
                }
            ]
        }
        emr_stub.add_response("describe_cluster", describe_cluster_response)
        emr_stub.add_response("list_instances", list_instances_response_page_1)
        emr_stub.add_response("list_instances", list_instances_response_page_2)
        emr_cluster = EMRCluster(cluster_id="j-3DD9ZR01DAU14", emr=emr)

        assert emr_cluster._cluster == describe_cluster_response["Cluster"]
        assert (
            emr_cluster._instances
            == list_instances_response_page_1["Instances"]
            + list_instances_response_page_2["Instances"]
        )
        assert emr_cluster._sec_conf is None
        assert not emr_cluster.is_krb_cluster


def test_get_cluster_happy_case_kerberos():
    emr = boto3.client("emr", region_name="us-west-2")
    with Stubber(emr) as emr_stub:
        describe_cluster_response = {
            "Cluster": {
                "Id": "j-3DD9ZR01DAU14",
                "Name": "Mycluster",
                "SecurityConfiguration": "kerb-security-config",
                "MasterPublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
            }
        }
        list_instances_response = {
            "Instances": [
                {
                    "Id": "j-3DD9ZR01DAU14",
                    "Ec2InstanceId": "i-0736242069217a485",
                    "PublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
                    "PublicIpAddress": "34.222.47.14",
                    "PrivateDnsName": "ip-172-31-1-113.us-west-2.compute.internal",
                    "PrivateIpAddress": "172.31.1.113",
                }
            ]
        }
        describe_sec_conf_response = {
            "Name": "kerb-security-config",
            "SecurityConfiguration": "{}",
        }
        emr_stub.add_response("describe_cluster", describe_cluster_response)
        emr_stub.add_response("list_instances", list_instances_response)
        emr_stub.add_response(
            "describe_security_configuration", describe_sec_conf_response
        )
        emr_cluster = EMRCluster(cluster_id="j-3DD9ZR01DAU14", emr=emr)

        assert emr_cluster._cluster == describe_cluster_response["Cluster"]
        assert emr_cluster._instances == list_instances_response["Instances"]
        assert emr_cluster._sec_conf == {}


def test_get_bad_cluster():
    emr = boto3.client("emr", region_name="us-west-2")
    with Stubber(emr) as emr_stub:
        list_instances_response = {
            "Instances": [
                {
                    "Id": "j-3DD9ZR01DAU14",
                }
            ]
        }
        emr_stub.add_client_error("describe_cluster", service_error_code=400)
        emr_stub.add_response("list_instances", list_instances_response)
        with pytest.raises(ValueError) as e:
            EMRCluster(cluster_id="j-3DD9ZR01DAU14", emr=emr)


@patch("sagemaker_studio_sparkmagic_lib.utils.get_domain_search")
def test_cluster_dedicated_krb_cluster(mock_utils):
    mock_utils.return_value = "test.me"
    emr = boto3.client("emr", region_name="us-west-2")
    with Stubber(emr) as emr_stub:
        describe_cluster_response = {
            "Cluster": {
                "Id": "j-3DD9ZR01DAU14",
                "Name": "Mycluster",
                "SecurityConfiguration": "kerb-security-config",
                "KerberosAttributes": {
                    "Realm": "KTEST.COM",
                    "KdcAdminPassword": "********",
                },
                "MasterPublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
            }
        }
        list_instances_response = {
            "Instances": [
                {
                    "Id": "j-3DD9ZR01DAU14",
                    "Ec2InstanceId": "i-0736242069217a485",
                    "PublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
                    "PublicIpAddress": "34.222.47.14",
                    "PrivateDnsName": "ip-172-31-1-113.us-west-2.compute.internal",
                    "PrivateIpAddress": "172.31.1.113",
                }
            ]
        }
        describe_sec_conf_response = {
            "Name": "kerb-security-config",
            "SecurityConfiguration": '{"EncryptionConfiguration": {"EnableInTransitEncryption": false, '
            '"EnableAtRestEncryption": false},"AuthenticationConfiguration": {'
            '"KerberosConfiguration": {"Provider": "ClusterDedicatedKdc", '
            '"ClusterDedicatedKdcConfiguration": {"TicketLifetimeInHours": 24 }}}}',
        }
        emr_stub.add_response("describe_cluster", describe_cluster_response)
        emr_stub.add_response("list_instances", list_instances_response)
        emr_stub.add_response(
            "describe_security_configuration", describe_sec_conf_response
        )
        emr_cluster = EMRCluster(cluster_id="j-3DD9ZR01DAU14", emr=emr)
        assert emr_cluster.is_krb_cluster
        assert (
            emr_cluster.primary_node_private_dns_name()
            == "ip-172-31-1-113.us-west-2.compute.internal"
        )
        assert (
            emr_cluster.primary_node_public_dns_name()
            == "ec2-34-222-47-14.us-west-2.compute.amazonaws.com"
        )
        krb_props = {
            "realms": {
                "KTEST.COM": {
                    "kdc": "ip-172-31-1-113.us-west-2.compute.internal:88",
                    "admin_server": "ip-172-31-1-113.us-west-2.compute.internal:749",
                    "default_domain": "test.me",
                }
            },
            "libdefaults": {"default_realm": "KTEST.COM", "ticket_lifetime": "24h"},
            "domain_realm": {
                "test.me": "KTEST.COM",
                ".test.me": "KTEST.COM",
            },
        }
        assert emr_cluster.get_krb_conf() == krb_props
        assert emr_cluster.get_kinit_user_name("ec2-user") == "ec2-user"


@patch("sagemaker_studio_sparkmagic_lib.utils.get_domain_search")
def test_cross_realm_krb_cluster(mock_utils):
    mock_utils.return_value = "test.me"
    emr = boto3.client("emr", region_name="us-west-2")
    with Stubber(emr) as emr_stub:
        describe_cluster_response = {
            "Cluster": {
                "Id": "j-3DD9ZR01DAU14",
                "Name": "Mycluster",
                "SecurityConfiguration": "kerb-security-config",
                "KerberosAttributes": {
                    "Realm": "EC2.INTERNAL",
                    "KdcAdminPassword": "********",
                },
                "MasterPublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
            }
        }
        list_instances_response = {
            "Instances": [
                {
                    "Id": "j-3DD9ZR01DAU14",
                    "Ec2InstanceId": "i-0736242069217a485",
                    "PublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
                    "PublicIpAddress": "34.222.47.14",
                    "PrivateDnsName": "ip-172-31-1-113.us-west-2.compute.internal",
                    "PrivateIpAddress": "172.31.1.113",
                }
            ]
        }
        describe_sec_conf_response = {
            "Name": "kerb-security-config",
            "SecurityConfiguration": '{"EncryptionConfiguration": {"EnableInTransitEncryption": false, '
            '"EnableAtRestEncryption": false},"AuthenticationConfiguration": {'
            '"KerberosConfiguration": {"Provider": "ClusterDedicatedKdc", '
            '"ClusterDedicatedKdcConfiguration": {"TicketLifetimeInHours": 24,'
            '"CrossRealmTrustConfiguration": {"Realm": "KOLLOJUN.NET","Domain": "kollojun.net",'
            ' "AdminServer": "kollojun.net", "KdcServer": "kollojun.net"}}}}}',
        }
        emr_stub.add_response("describe_cluster", describe_cluster_response)
        emr_stub.add_response("list_instances", list_instances_response)
        emr_stub.add_response(
            "describe_security_configuration", describe_sec_conf_response
        )
        emr_cluster = EMRCluster(cluster_id="j-3DD9ZR01DAU14", emr=emr)
        assert emr_cluster.is_krb_cluster
        assert (
            emr_cluster.primary_node_private_dns_name()
            == "ip-172-31-1-113.us-west-2.compute.internal"
        )
        assert (
            emr_cluster.primary_node_public_dns_name()
            == "ec2-34-222-47-14.us-west-2.compute.amazonaws.com"
        )
        krb_props = {
            "realms": {
                "EC2.INTERNAL": {
                    "kdc": "ip-172-31-1-113.us-west-2.compute.internal:88",
                    "admin_server": "ip-172-31-1-113.us-west-2.compute.internal:749",
                    "default_domain": "test.me",
                },
                "KOLLOJUN.NET": {
                    "kdc": "kollojun.net",
                    "admin_server": "kollojun.net",
                    "default_domain": "kollojun.net",
                },
            },
            "libdefaults": {"default_realm": "EC2.INTERNAL", "ticket_lifetime": "24h"},
            "domain_realm": {
                "test.me": "EC2.INTERNAL",
                ".test.me": "EC2.INTERNAL",
                ".kollojun.net": "KOLLOJUN.NET",
                "kollojun.net": "KOLLOJUN.NET",
            },
        }
        assert emr_cluster.get_krb_conf() == krb_props
        assert emr_cluster.get_kinit_user_name("ec2-user") == "ec2-user"


@patch("sagemaker_studio_sparkmagic_lib.utils.get_domain_search")
def test_external_kdc_cluster(mock_utils):
    mock_utils.return_value = "test.me"
    emr = boto3.client("emr", region_name="us-west-2")
    with Stubber(emr) as emr_stub:
        describe_cluster_response = {
            "Cluster": {
                "Id": "j-3DD9ZR01DAU14",
                "Name": "Mycluster",
                "SecurityConfiguration": "kerb-security-config",
                "KerberosAttributes": {
                    "ADDomainJoinPassword": "********",
                    "Realm": "MYKERBEROSREALM.COM",
                    "ADDomainJoinUser": "********",
                    "KdcAdminPassword": "********",
                },
                "MasterPublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
            }
        }
        list_instances_response = {
            "Instances": [
                {
                    "Id": "j-3DD9ZR01DAU14",
                    "Ec2InstanceId": "i-0736242069217a485",
                    "PublicDnsName": "ec2-34-222-47-14.us-west-2.compute.amazonaws.com",
                    "PublicIpAddress": "34.222.47.14",
                    "PrivateDnsName": "ip-172-31-1-113.us-west-2.compute.internal",
                    "PrivateIpAddress": "172.31.1.113",
                }
            ]
        }
        describe_sec_conf_response = {
            "Name": "kerb-security-config",
            "SecurityConfiguration": '{"AuthenticationConfiguration":{"KerberosConfiguration":'
            '{"ExternalKdcConfiguration":{"KdcServerType":"Single","AdIntegrationConfiguration":'
            '{"AdRealm":"MYADDOMAIN.COM","AdDomain":"myaddomain.com","AdServer":"myaddomain.com"},'
            '"AdminServer":"ip-10-0-0-164.us-west-2.compute.internal",'
            '"KdcServer":"ip-10-0-0-164.us-west-2.compute.internal"}'
            ',"Provider":"ExternalKdc"}}}',
        }
        emr_stub.add_response("describe_cluster", describe_cluster_response)
        emr_stub.add_response("list_instances", list_instances_response)
        emr_stub.add_response(
            "describe_security_configuration", describe_sec_conf_response
        )
        emr_cluster = EMRCluster(cluster_id="j-3DD9ZR01DAU14", emr=emr)
        assert emr_cluster.is_krb_cluster
        assert (
            emr_cluster.primary_node_private_dns_name()
            == "ip-172-31-1-113.us-west-2.compute.internal"
        )
        assert (
            emr_cluster.primary_node_public_dns_name()
            == "ec2-34-222-47-14.us-west-2.compute.amazonaws.com"
        )
        krb_props = {
            "realms": {
                "MYKERBEROSREALM.COM": {
                    "kdc": "ip-10-0-0-164.us-west-2.compute.internal",
                    "admin_server": "ip-10-0-0-164.us-west-2.compute.internal",
                    "default_domain": "test.me",
                },
                "MYADDOMAIN.COM": {
                    "kdc": "myaddomain.com",
                    "admin_server": "myaddomain.com",
                    "default_domain": "myaddomain.com",
                },
            },
            "libdefaults": {
                "default_realm": "MYKERBEROSREALM.COM",
                "ticket_lifetime": "24h",
            },
            "domain_realm": {
                "test.me": "MYKERBEROSREALM.COM",
                ".test.me": "MYKERBEROSREALM.COM",
                ".myaddomain.com": "MYADDOMAIN.COM",
                "myaddomain.com": "MYADDOMAIN.COM",
            },
        }
        assert emr_cluster.get_krb_conf() == krb_props
        assert emr_cluster.get_kinit_user_name("ec2-user") == "ec2-user@MYADDOMAIN.COM"
        assert emr_cluster.krb_hostname_override() == "ip-172-31-1-113.test.me"


sagemaker_studio_sparkmagic_lib/emr.py
import json
import os
import sys

import boto3
import botocore
import logging
from sagemaker_studio_sparkmagic_lib import utils

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)
logger.addHandler(logging.StreamHandler(sys.stdout))


class EMRCluster:
    """
    class that represents emr cluster and provides helper functions to get relevant attributes
    needed to generate configurations
    """

    def __init__(self, cluster_id=None, role_arn=None, emr=None):
        self._cluster = None
        self._instances = None
        self._sec_conf = None
        self._emr_primary_instance = None

        if not emr:
            sess = self._get_boto3_session(role_arn)
            emr = sess.client(
                "emr", endpoint_url=utils.get_emr_endpoint_url(self._get_region())
            )
        self._get_cluster(emr, cluster_id)
        self._get_instances(emr, cluster_id)
        self._get_security_conf(emr)
        logger.info(f"Successfully read emr cluster({cluster_id}) details")
        self._identify_emr_primary_instance()

    def _get_boto3_session(self, role_arn):
        """
        Allows cross account cluster access
        """
        if not role_arn:
            return boto3.session.Session()
        else:
            logger.info(f"Assuming role: {role_arn}")
            sts_client = boto3.client("sts")
            try:
                assume_role_object = sts_client.assume_role(
                    RoleArn=role_arn, RoleSessionName="SageMakerStudioUser"
                )
            except botocore.exceptions.ClientError as ce:
                logger.debug(
                    f"Failed to assume role: ({role_arn}) details. {ce.response}"
                )
                raise ValueError(
                    f"Unable to assume role(arn: {role_arn}). Ensure permissions are setup correctly"
                    f' Error: {ce.response["Error"]}'
                ) from None
            return boto3.Session(
                aws_access_key_id=assume_role_object["Credentials"]["AccessKeyId"],
                aws_secret_access_key=assume_role_object["Credentials"][
                    "SecretAccessKey"
                ],
                aws_session_token=assume_role_object["Credentials"]["SessionToken"],
            )

    def _get_cluster(self, emr, cluster_id):
        try:
            describe_cluster_response = emr.describe_cluster(ClusterId=cluster_id)
        except botocore.exceptions.ClientError as ce:
            logger.debug(
                f"Failed to get EMR cluster({cluster_id}) details. {ce.response}"
            )
            raise ValueError(
                f"Unable to get Emr Cluster(Id: {cluster_id}) details using describe-cluster API."
                f' Error: {ce.response["Error"]}'
            ) from None
        logger.debug(f"Describe emr cluster response: {describe_cluster_response}")
        self._cluster = describe_cluster_response["Cluster"]

    def _get_instances(self, emr, cluster_id):
        try:
            paginator = emr.get_paginator("list_instances")
            page_iterator = paginator.paginate(ClusterId=cluster_id)
            instances = []
            for page in page_iterator:
                instances.extend(page["Instances"])
        except botocore.exceptions.ClientError as ce:
            logger.debug(
                f"Failed to list instances in  EMR cluster({cluster_id}) details. {ce.response}"
            )
            raise ValueError(
                f"Unable to list instances in EMR Cluster(Id: {cluster_id}) details using list-instances API."
                f' Error: {ce.response["Error"]}'
            ) from None
        logger.debug(f"List instances response: {instances}")
        self._instances = instances

    def _identify_emr_primary_instance(self):
        """
        Identifies EMR primary node.
        """
        # Primary node public dns can actually be private DNS if EMR cluster is launched in
        # private subnet hence we check both names. See
        # https://docs.aws.amazon.com/emr/latest/APIReference/API_Cluster.html
        primary_node_public_dns = self._cluster["MasterPublicDnsName"]
        for instance in self._instances:
            if (
                "PublicDnsName" in instance
                and instance["PublicDnsName"] == primary_node_public_dns
            ) or (
                "PrivateDnsName" in instance
                and instance["PrivateDnsName"] == primary_node_public_dns
            ):
                self._emr_primary_instance = instance

        if self._emr_primary_instance is None:
            raise ValueError(
                f"Failed to find primary node ip address needed to communicate with Livy server."
                f"Please ensure clusterId{self._cluster['Id']} is correct"
            ) from None

    def _get_security_conf(self, emr):
        if "SecurityConfiguration" not in self._cluster:
            logger.debug(
                "Skipping describing security group as no specific security configuration is used for cluster"
            )
            self._sec_conf = None
            return
        security_conf = self._cluster["SecurityConfiguration"]
        cluster_id = self._cluster["Id"]
        try:
            describe_sec_conf_response = emr.describe_security_configuration(
                Name=security_conf
            )
        except botocore.exceptions.ClientError as ce:
            logger.debug(
                f"Failed to get security configuration details({security_conf}) of EMR cluster({cluster_id})"
                f"details. {ce.response}"
            )
            raise ValueError(
                f"Unable to get security configuration details({security_conf}) of EMR CLuster(Id: {cluster_id})"
                f'details using describe-security-configuration. Error: {ce.response["Error"]}'
            ) from None
        logger.debug(
            f"Describe emr security config response: {describe_sec_conf_response}"
        )
        self._sec_conf = json.loads(describe_sec_conf_response["SecurityConfiguration"])

    @property
    def is_krb_cluster(self):
        """
        Returns true if EMR cluster is configured with kerberos
        """
        return not (
            "KerberosAttributes" not in self._cluster
            or len(self._cluster["KerberosAttributes"]) == 0
        )

    def primary_node_private_dns_name(self):
        """
        Returns primary node private DNS name
        """
        return self._emr_primary_instance["PrivateDnsName"]

    def primary_node_public_dns_name(self):
        """
        Returns primary node public DNS name
        """
        return self._emr_primary_instance["PublicDnsName"]

    def krb_hostname_override(self):
        """
        According to kerberos https://github.com/requests/requests-kerberos/blob/master/README.rst
        "
        If communicating with a host whose DNS name doesn't match its kerberos hostname (eg, behind a content switch or
        load balancer), the hostname used for the Kerberos GSS exchange can be overridden by setting the hostname_override
        "

        Since we are communicating from studio host not registered in EMR one has to override hostname matching DNS/DHCP options
        """
        hostname = self.primary_node_private_dns_name()
        search = utils.get_domain_search(self._get_region())
        if search != utils.get_default_domain_search(self._get_region()):
            names = hostname.split(".")
            if len(names) > 1:
                hostname = f"{names[0]}.{search}"

        return hostname

    def get_kinit_user_name(self, user_name):
        """
        Generates exact user name to be used for kinit. Depending on Kerberos configuration user may
        have to use realm name to do kinit
        """
        if self.is_krb_cluster:
            sec_krb_conf = self._sec_conf["AuthenticationConfiguration"][
                "KerberosConfiguration"
            ]
            krb_provider = sec_krb_conf["Provider"]
            if krb_provider == "ExternalKdc":
                kdc_conf = sec_krb_conf["ExternalKdcConfiguration"]
                ad_integ_conf = kdc_conf["AdIntegrationConfiguration"]
                return f"{user_name}@{ad_integ_conf['AdRealm']}"

        return user_name

    def is_external_kdc(self):
        if not self.is_krb_cluster:
            return False
        sec_krb_conf = self._sec_conf["AuthenticationConfiguration"][
            "KerberosConfiguration"
        ]
        krb_provider = sec_krb_conf["Provider"]

        return krb_provider == "ExternalKdc"

    def get_krb_conf(self):
        """
        Generate kerberos configuration parameters for a given EMR cluster.
        These configuration parameters are used for constructing krb5.conf
        Returned configuration is a two layered dictionary with top level keys for "sections"
        defined in kerberos config
        https://web.mit.edu/kerberos/krb5-1.12/doc/admin/conf_files/krb5_conf.html#sections
        {
            "libdefaults": {}
            "realms": {}
            "domain_realm": {}
        }
        """
        # Good overview of kerberos configuration
        # https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-create-security-configuration.html#emr-kerberos-cli-parameters
        # Kerberos server config on EMR cluster
        properties = {}
        emr_realm_name = self._cluster["KerberosAttributes"]["Realm"]
        # DNS Search value is used for domain_realm mappings and other kerberos properties
        # We have learned that using DNS Search is needed instead of AWS default DNS values for kerberos to work
        search = utils.get_domain_search(self._get_region())

        emr_realm_attr = {
            "kdc": f"{self.primary_node_private_dns_name()}:88",
            "admin_server": f"{self.primary_node_private_dns_name()}:749",
            "default_domain": f"{search}",
        }

        properties["realms"] = {emr_realm_name: emr_realm_attr}
        properties["libdefaults"] = {"default_realm": emr_realm_name}

        # map emr realm to domain search value
        properties["domain_realm"] = {
            search: f"{emr_realm_name}",
            # not the same key as above. There is a dot at the beginning of key
            f".{search}": f"{emr_realm_name}",
        }
        sec_krb_conf = self._sec_conf["AuthenticationConfiguration"][
            "KerberosConfiguration"
        ]
        krb_provider = sec_krb_conf["Provider"]
        if krb_provider == "ClusterDedicatedKdc":
            kdc_conf = sec_krb_conf["ClusterDedicatedKdcConfiguration"]
            properties["libdefaults"][
                "ticket_lifetime"
            ] = f'{kdc_conf.get("TicketLifetimeInHours", "24")}h'
            if "CrossRealmTrustConfiguration" in kdc_conf:
                cross_real_conf = kdc_conf["CrossRealmTrustConfiguration"]
                cross_realm = cross_real_conf["Realm"]
                cross_domain = cross_real_conf["Domain"]
                properties["realms"][cross_real_conf["Realm"]] = {
                    "kdc": cross_real_conf["KdcServer"],
                    "admin_server": cross_real_conf["AdminServer"],
                    "default_domain": cross_real_conf["Domain"],
                }
                properties["domain_realm"][cross_domain] = cross_realm
                properties["domain_realm"][f".{cross_domain}"] = cross_realm
        elif krb_provider == "ExternalKdc":
            kdc_conf = sec_krb_conf["ExternalKdcConfiguration"]
            properties["libdefaults"][
                "ticket_lifetime"
            ] = f'{kdc_conf.get("TicketLifetimeInHours", "24")}h'

            # For external kdc configuration default realm-properties should point to external kdc server
            emr_realm_attr = {
                "kdc": kdc_conf["KdcServer"],
                "admin_server": kdc_conf["AdminServer"],
                "default_domain": f"{search}",
            }

            properties["realms"][emr_realm_name] = emr_realm_attr
            ad_integ_conf = kdc_conf["AdIntegrationConfiguration"]
            properties["realms"][ad_integ_conf["AdRealm"]] = {
                # For external kdc configuration parameter AdServer is not documented in EMR public docs
                # We found from some use cases that parameter AdServer Exists
                # and it should be used as KDC Server
                "kdc": ad_integ_conf.get("AdServer", ad_integ_conf["AdDomain"]),
                "admin_server": ad_integ_conf.get(
                    "AdServer", ad_integ_conf["AdDomain"]
                ),
                "default_domain": ad_integ_conf["AdDomain"],
            }
            ext_realm = ad_integ_conf["AdRealm"]
            ext_domain = ad_integ_conf["AdDomain"]
            properties["domain_realm"][ext_domain] = ext_realm
            properties["domain_realm"][f".{ext_domain}"] = ext_realm

        return properties

    def _get_region(self):
        return os.getenv("AWS_REGION", "us-west-2")
